CreeDictionary/tests/test_docs.py

```diff
@@ -0,0 +1,146 @@
+from pathlib import Path
+from gitignore_parser import parse_gitignore
+from os import path
+import os
+from typing import List, Callable, NoReturn
+import toml
+import pytest
+from sys import stderr
+from functools import partial
+from sortedcontainers import SortedSet
+
+IgnoreCheck = Callable[[Path], bool]
+"""
+If a path is ignored, it gives True. If a path is not ignored, it gives False
+"""
+
+FILE_DIR = Path(path.dirname(__file__))
+
+
+def get_project_root():
+    """
+    bubbles up until we see a directory with .git folder
+
+    :return: The root directory of the repository
+    """
+    # Uhh I actually don't know any reliable way to do this
+    # this sounds like a reasonable method. You are Welcome to fix it.
+    current = FILE_DIR
+    while not (current / ".git").is_dir():
+        current = current.parent
+    return current
+
+
+PROJECT_ROOT = get_project_root()
+
+
+def _get_not_ignored_dirs(root: Path, ancestral_ignores: List[IgnoreCheck]) -> List[Path]:
+    """
+    :param root: must be a existent directory
+    :return: all dirs that are not gitignored, not including root. sorted in a depth first manner
+    """
+    # .is_dir also checks existence
+    assert root.is_dir(), "dumb dumb: root must be a directory :broken_heart:"
+
+    ret = []
+
+    ignores_in_control = ancestral_ignores.copy()
+
+    # is_file also checks for existence
+    for own_ignore in filter(Path.is_file, [root / ".gitignore", root / ".dirdocignore"]):
+        ignores_in_control.append(parse_gitignore(own_ignore))
+
+    # glob('*') returns direct children only
+    for child_dir in filter(Path.is_dir, root.glob("*")):
+        # for/else is used by intention here for early return pattern
+        for ignore in ignores_in_control:
+            if ignore(child_dir):
+                break
+        else:  # not ignored
+            ret.append(child_dir)
+            ret.extend(_get_not_ignored_dirs(child_dir, ignores_in_control))
+
+    return ret
+
+
+def get_not_ignored_dirs() -> List[Path]:
+    """
+    get all directories in this project that are not included in .gitignore or .dirdocignore
+    """
+    return _get_not_ignored_dirs(PROJECT_ROOT, [])
+
+
+def fail_and_provide_suggestion(reason: str, suggestion: str) -> NoReturn:
+    print(reason, file=stderr)
+    print(suggestion, file=stderr)
+    pytest.fail(reason)
+    raise Exception("If you see this. Some thing impossible has happened. This line shouldn't be reached.")
```

Comment:
Have to add this line, otherwise mypy screams "This function has an implicit return but is marked as NoReturn"